core/src/main/java/forklift/consumer/Consumer.java

```diff
@@ -483,10 +405,154 @@ else if (producer.topic().length() > 0)
         return closeMe;
     }
 
+    private Object constructMessageHandlerInstance(ForkliftMessage forkliftMessage, List<Closeable> closeables) throws IllegalAccessException, InvocationTargetException, InstantiationException, IOException {
+        Object instance = null;
+        if (this.constructor != null) {
+            Object[] constructorParameters = buildConstructorParameters(forkliftMessage, closeables);
+            instance = this.constructor.newInstance(constructorParameters);
+        } else {
+            instance = msgHandler.newInstance();
+        }
+        return instance;
+    }
+
+    private Object[] buildConstructorParameters(ForkliftMessage forkliftMessage, List<Closeable> closeables) throws IOException {
+        Object[] parameters = new Object[constructorAnnotations.length];
+        int index = 0;
+        for (Annotation[] parameterAnnotations : constructorAnnotations) {
+            //there must be at least one annotation that indicates what we should inject
+            Annotation injectable = null;
+            for (Annotation parameterAnnotation : parameterAnnotations) {
+                if (injectFields.containsKey(parameterAnnotation.annotationType())) {
+                    injectable = parameterAnnotation;
+                    break;
+                }
+            }
+            Parameter p = constructor.getParameters()[index];
+            Object value = getInjectableValue(injectable, null, p.getType(), forkliftMessage);
+            parameters[index] = value;
+            if (value != null && value instanceof ForkliftProducerI) {
+                closeables.add((ForkliftProducerI)value);
+            }
+
+            index++;
+        }
+        return parameters;
+    }
+
+    private Object getInjectableValue(Annotation decorator, String mappedName, Class<?> mappedClass, ForkliftMessage msg) throws IOException {
```

Comment:
Thought for later: maybe this could be separated out better with a map of handlers or something